text/0000-pub-macro-rules.md

```diff
@@ -0,0 +1,300 @@
+    * Annotate the invocation with the path to the parent module where the macro is defined. 
+* Remove `#[macro_export]` annotations from macros
+    * mark macros as `pub`
+    * publicly re-export them from the top-level module of the crate where the macro is defined
+* Remove `#[macro_use]` where annotating a module
+    * mark all the macros in the previously annotated module as `pub(crate)`
+    * mark the module itself as pub(crate) if not already `pub(crate)` or `pub`
+    * annotate the macro invocation with a path to the module where the macro is defined.
+* Remove `#[macro_use]` from from `extern crate` item
+    * change all uses of macros to qualified `$EXTERNAL_CRATE::$MACRO_NAME` invocations. 
+
+These steps should be automatable so that rustfix can be used to aid in migration though there are sufficiently complex use cases that a fully automatable transition is likely to not be possible. For example, the "anonymous" helper macros use case discussed in the section on shadowing would not be able to fixed in an automated way.  
+
+Roughly what percentage of use cases will be machine migratable is an open question. 
+
+## Translating Common Patterns 
+
+The following are how common patterns in macros today translate to the next path based scoping system.
+
+### Deeply nested macros
+
+Macro use makes all macros inside a child module available to the parent module.
+
+```rust 
+
+#[macro_use]
+mod m {
+    #[macro_use]
+    mod n {
+        macro_rules! define_foo {  () => { fn foo() {} } }
+    }
+}
+    
+    
+define_foo!();
+```
+
+This would be translated as: 
+
+```rust 
+pub(crate) mod m {
+    pub(crate) mod n {
+        pub(crate) macro_rules! define_foo {  () => { fn foo() {} } }
+    }
+}
+    
+    
+m::n::define_foo!();
+```
+
+### Recursive macros
+
+Recursive macros are macros that call themselves (perhaps with different arguments)
+
+```rust 
+#[macro_use]
+mod m {
+    macro_rules! print_expr {
+        ($e:expr) => {{
+            println!("Going to do {}", stringify!($e));
+            print_expr!(no_print => $e)
+        }};
+        (no_print => $e:expr) => {{
+            $e
+        }};
+    }
+}
+
+fn main() {
+    print_expr!(1 + 1)
+}
+```
+
+Naively changing this to path based scope would not work as it is not guaranteed that the unqualified `print_expr` name is in scope. In the example above, `print_expr!` is used recursively inside the macro, but in a path scoped system the recursive call would not be in scope if the macro was called with a qualified path (e.g., user calls `m::print_expr!` which references unqualified `print_expr!` which is not in scope).
+
+A possible way to handle this is to introduce a new macro specific keyword `$self` which is directly analogous to `$crate` except that it refers to the module where the macro is defined. This would work in the simple case but quickly breaks down in more complicated module paths. For example:
```

Comment:
There is also the question of macros that would use `$self` as a capture; but I don't think `rustfix` would encounter a recursive and `$self`-binding macro very often 🙃